### DISCUSSION TITLE
Understanding the Timing of .emit() After .join() in Socket.IO

### DISCUSSION THREAD

**Discussion by KishonShrill** [socketio/socket.io#5015] [Q&A]
Is it feasible to invoke the .emit() method immediately after calling the .join() method in Socket.IO? 

While developing a Rock-Paper-Scissors game utilizing Socket.IO for real-time communication, I encountered an issue wherein newly joined players were not receiving notification messages upon joining a room. My intention was to promptly inform the joining player about their opponent's userID and display a "Match has been found" message. However, despite emitting the event within the join-room event listener, the notification was successfully dispatched to the opponent, while the newly joined player did not receive the opponent's userID or the match notification.

The relevant snippet of the code is as follows:
```javascript
clientSocket.on('join-room', (roomToJoin, player) => {
    try {
      if (roomToJoin.length !== 10) {
        clientSocket.emit('room-not-valid', "Room does not exist...")
        throw new Error("Key and value must be exactly 10 characters long.")
      }
      if (rooms[roomToJoin].length >= MAX_VALUE_PER_KEY) {
        clientSocket.emit('room-already-full', `Maximum limit (${MAX_VALUE_PER_KEY}) reached for key '${roomToJoin}'`)
        throw new Error("Room is already full!!!")
      }

      // Join a room
      createRoom(roomToJoin, player) 
      clientSocket.join(roomToJoin) /* - - - - USER WILL JOIN HERE - - - - */
      console.log(`${player} has made a room on ${roomToJoin}`)
      console.log({rooms})

      if (rooms[roomToJoin].length >= MAX_VALUE_PER_KEY) {
        const existingPlayer = rooms[roomToJoin][0];
        clientSocket.to(roomToJoin).emit('match-found', [existingPlayer, player]); /* - - - - EMIT IS IN HERE - - - - */
        console.log("I HAVE SUCCESSFULLY CONNECTED YAHOO")
      }
    } catch (error) {
      console.error(error)
    }
```
`rooms[roomToJoin].length` is true to `MAX_VALUE_PER_KEY`

What could be the underlying reason for a newly joined socket failing to receive an emit event, despite the .emit() function being called after the .join() operation? Additionally, how can this issue be addressed to ensure that all players in the room receive the appropriate notifications promptly upon joining?

From what I understood while experimenting around the socket, it WILL receive the `.emit` function as long as it already has `.join` on the start, but won't if it has just joined recently...

For the complete code and further reference, please visit the following GitHub repository: [Rock-Paper-Scissors Game Repository](https://github.com/KishonShrill/rock-paper-scissor)

**Comment by darrachequesne** [comment]
Hi!

> What could be the underlying reason for a newly joined socket failing to receive an emit event, despite the .emit() function being called after the .join() operation? Additionally, how can this issue be addressed to ensure that all players in the room receive the appropriate notifications promptly upon joining?

Hmm, this should not happen. Are you sure the client is listening to the 'match-found' event?

**Reply by KishonShrill** [reply]
Yes, a matter of fact, the emit will only be sent to players who are already inside the room but wont send the emit on the recently joined player
